File: Laboratory 3/lab3.ml
(* HELPER FUNCTIONS *)

let rec stringToCharList s =
  let rec helper (s, it) =
    match String.get s it with
    | exception (Invalid_argument s) -> []
    | _ -> String.get s it::helper(s, it+1)
  in helper(s,0)
;;    
let rec contains (list,elem) =
    match list with
    | [] -> false
    |head::tail -> if head = elem then true else contains(tail, elem)
;;


(* LAB 3  *)



let rec lab21 list =
    match list with
    | [] -> None
    | [head] -> Some(head)
    | head::tail -> lab21(tail)
  ;;

  lab21 [1;4;5;6];;
  lab21 [];;

let rec lab22 list =
    match list with
    | [] -> None
    | [head] -> None
    | [head1;head2] -> Some([head1; head2])
    | head1::head2::tail -> lab22(head2::tail)
  ;;

lab22 [];;
lab22 [3];;
lab22 [5;6];;
lab22 [6;8;2;4];;

let rec lab23 list =
  match list with
  | [] -> 0
  | head::tail -> lab23(tail) + 1
;;

lab23 [];;
lab23 [5;3;5];;

let rec lab24 list = 
  let rec helper (list,result) = 
    match list with
    | [] -> []
    | [head] -> head::result
    | head::tail -> helper(tail, result) @ [head]
  in helper(list, [])
;;

lab24 [];;
lab24 [5];;
lab24 [5;7;8];;
let lab25 s = 
  if stringToCharList s = lab24(stringToCharList s) then true else false;;

lab25 "cokolwiek";;
lab25 "kajak";;  
let rec lab26 list = 
  let rec helper(list, result) =
    match list with
    |[] -> result
    |head::tail -> if contains(tail, head) then helper(tail, result) else head::helper(tail, result)
  in helper(list, [])
;;

lab26 [1;2;3;4];;
lab26 [3;3;3;3];;
lab26 [6];;
lab26 [];;
lab26 [1;2;2;4;4;5;5;6;6];;

let rec lab27 list =
  let rec helper (list, counter, result) =
    if counter mod 2=0 then
      match list with
      | [] -> []
      | [head] -> head::result
      | head::tail -> head:: helper(tail, counter+1, result)
    else
      match list with
      | [] -> []
      | [head] -> result
      | head::tail -> helper(tail, counter+1, result)
    in helper(list, 0, [])    
  ;;

lab27 [];;
lab27 [1];;
lab27 [1;5;7;8];;
let rec lab28 number = 
  if number < 2 then false else
  if number = 2 then true else
  let rec helper (number, toDivideBy) =
      if toDivideBy = number then true else
      if number mod toDivideBy = 0 then false else helper(number, toDivideBy+1)
    in helper(number, 2)
  ;;  

lab28 1;;
lab28 8;;
lab28 11;;

File: Laboratory 3/lab3.scala
@main def main() =

  println("Zadanie 1: ")
  println(lab21(List(1,4,7,6)))
  println(lab21(List()))
  println()

  println("Zadanie 2: ")
  println(lab22(List()))
  println(lab22(List(3)))
  println(lab22(List(4,5)))
  println(lab22(List(1,2,3,4,5,6)))
  println()

  println("Zadanie 3: ")
  println(lab23(List()))
  println(lab23(List(1,2,3)))
  println()

  println("Zadanie 4: ")
  println(lab24(List()))
  println(lab24(List(5)))
  println(lab24(List(5,7,8)))
  println()

  println("Zadanie 5: ")
  println(lab25("kajak"))
  println(lab25("cokolwiek"))
  println()

  println("Zadanie 6: ")
  println(lab26(List()))
  println(lab26(List(1,2,3,4)))
  println(lab26(List(3,3,3,3)))
  println(lab26(List(6)))
  println(lab26(List(1,2,2,2,3,3,5,5,6,6)))
  println()

  println("Zadanie 7: ")
  println(lab27(List()))
  println(lab27(List(1)))
  println(lab27(List(1,5,7,8)))
  println()

  println("Zadanie 8:")
  println(lab28(1))
  println(lab28(11))
  println(lab28(6))
  println(lab28(17))


def lab21(list: List[Any]) : Option[Any] =
  list match {
    case Nil => None
    case head::Nil => Some(head)
    case _::tail => lab21(tail)
  }

def lab22(list: List[Any]) : Option[List[Any]] =
  list match{
    case Nil => None
    case head:: Nil => None
    case head:: elem2 :: Nil => Some(List(head, elem2))
    case head :: elem2 :: tail => lab22(elem2::tail)
  }

def lab23(list: List[Any]): Int =
  list match
    case Nil => 0
    case head :: tail => lab23(tail) + 1

def lab24(list: List[Any]) : List[Any] =
  def helper(list: List[Any],result: List[Any]) : List[Any] =
    list match
      case Nil => List()
      case head :: Nil => head :: result
      case head :: tail => helper(tail, result) ::: List(head)
  helper(list, List.empty)

def lab25(word: String) : Boolean =
  if word.toList == lab24(word.toList) then true else false

def lab26(list: List[Any]) : List[Any] =
  def helper(list: List[Any], result: List[Any]) : List[Any] =
    list match
      case Nil => result
      case head :: tail =>
        val result_temp = helper(tail, result)
        if !contains(tail, head) then head :: result_temp else result_temp
  helper(list, List.empty)

def lab27(list: List[Any]): List[Any] =
  def helper(list: List[Any], counter: Int, result: List[Any]) : List[Any] =
    if counter%2==0 then
      list match
        case Nil => List()
        case head :: Nil => head :: result
        case head:: tail => head :: helper(tail, counter+1, result)


    else
      list match
        case Nil => List()
        case head :: Nil => result
        case head :: tail => helper(tail, counter + 1, result)


  helper(list, 0, List.empty)

def lab28(number: Int) : Boolean =
  if number < 2 then return false
  if number == 2 then true
  else
    def helper(number: Int, toDivideBy: Int): Boolean =
      if toDivideBy == number then return true
      if number % toDivideBy == 0 then false else helper(number, toDivideBy + 1)
    helper(number, 2)

def contains(list: List[Any], elem: Any) : Boolean =
  list match
    case Nil => false
    case head :: tail => if head == elem then true else contains(tail, elem)

